books/std/util/defines.lisp

```diff
@@ -440,34 +449,59 @@ encapsulate), and is mainly meant as a tool for macro developers.</dd>
                        parents
                      (and cliquename
                           (not (eq cliquename guts.name))
-                          (list cliquename)))))
+                          (list cliquename))))
+       (divide-defsections (and process-returns divide-defsections))
+       (rest-events
+        `(
+; Matt K. mod, 8/27/2021, for GitHub Issue #1302: allow current-function to be
+; seen even when including an uncertified book, by making the following
+; non-local.
+          (set-define-current-function ,guts.name)
+          (with-output :stack :pop (progn . ,guts.rest-events)))))
 
-    `((defsection ,guts.name
+    (mv
+     `(defsection ,guts.name
         ,@(and parents `(:parents ,parents))
         ,@(and short   `(:short ,short))
         ,@(and long    `(:long ,long))
         ,@(and process-returns
                `((make-event
-                  (let* ((world (w state))
-                         (events (returnspec-thms ',guts.name
-                                                  ',guts.name-fn
-                                                  ',guts.returnspecs
-                                                  world)))
-                    `(with-output :stack :pop (progn . ,events))))))
-; Matt K. mod, 8/27/2021, for GitHub Issue #1302: allow current-function to be
-; seen even when including an uncertified book, by making the following
-; non-local.
-        (set-define-current-function ,guts.name)
-        (with-output :stack :pop (progn . ,guts.rest-events)))
-      ;; Make sure the section gets processed first.  Once it's done,
-      ;; we can add the signature block.
-      (with-output :on (error) ,(add-signature-from-guts guts)))))
-
-(defun collect-fn-defsections (gutslist cliquename process-returns)
+                   (let* ((world (w state))
+                          (events (returnspec-thms ',guts.name
+                                                   ',guts.name-fn
+                                                   ',guts.returnspecs
+                                                   world)))
+                     `(with-output :stack :pop (progn . ,events))))))
+        ,@(and (not divide-defsections)
+               rest-events))
+     `(,@(and divide-defsections
```

Comment:
Do we have to divide it into two defsections to support putting verify-guards as a separate event?  Why not insert the verify-guards into the one defsection, between the returnspec-thms and the rest-events?

Reply:
I think we need a single `verify-guards` after all of the returnspec-thms (from all the functions in the `defines`) but before their collective rest-events.

To be concrete, say we have a `defines` likes this:
```lisp
(defines foo
  :returns-no-induct t
  :verify-guards :after-returns
  (define foo (...)
    ...
    ///
    foo-other-events)
  (define bar (...)
    ...
    ///
    bar-other-events))
```
If we don't split the defsections, we'll generate:
```lisp
...
(defsection foo
  ...
  foo-return-thms
  foo-other-events
  ...)
(defsection bar
  ...
  bar-return-thms
  bar-other-events
  ...)
...
```
We want to put the `(verify-guards foo)` (verifying guards for both `foo` and `bar`) after `foo-return-thms` and `bar-return-thms`, but before `foo-other-events` and `bar-other-events`. With the current order of things, we can't do that. Hence why I thought we should generate (just when `:returns-no-induct t` and `:verify-guards :after-returns`):
```lisp
(defsection foo
  ...
  foo-return-thms)
(defsection bar
  ...
  bar-return-thms)
(verify-guards foo)
(defsection foo
  :extension foo
  foo-other-events
  ...)
(defsection bar
  :extension bar
  bar-other-events
  ...)
```

Reply:
I see, that makes sense. How about additionally checking whether the function has rest-events at all before producing a second defsection for it? I think it might be fairly rare for the individual functions in the clique to have rest-events.

Reply:
Oh good point. I pushed a change to only divide sections if there are rest-events.